### ISSUE TITLE
Sometimes "License Information" field is empty

### ISSUE THREAD

**Issue by leimaohui** [fossology/fossology-python#20]
Sometimes, especially when invoke multi fossology-python at the same time, "License Information" field is empty without error or warning.The spdx file look like as following:
```shell
$ cat  build/tmp/deploy/spdx/m4-1.4.18.spdx | tail
© Seindal seindal@diku.dk Richard A Nelson cowboy@vnet.ibm.com Richard Stallman rms@gnu.org Robert Bernstein rocky@panix.com Robert Bihlmeyer robbe@orcus.priv.at Roland H. Pesch roland@wrs.com Roland McGrath roland@gnu.org Ronny Peine RonnyPeine@gmx.de Sami Liedes sliedes@cc.hut.fi Santiago Vila sanvila@debian.org Scott Bartram deneb!scottb Scott McCreary scottmc2@gmail.com Sergey Poznyakoff gray@gnu.org.ua Simon Leinen simon@lia.di.epfl.ch Skip Montanaro skip@automatrix.com Stepan Kasal kasal@ucw.cz Stephen Perkins perkins@cps.msu.edu Steve Williamson willy@uinpla.npl.uiuc.edu Steven Augart saugart@yahoo.com Thomas Klausner tk@giga.or.at Thorsten Ohl ohl@physics.harvard.edu Todd Rinaldo toddr@cpanel.net Tom G. Christensen tgc@jupiterrise.com Tom McConnell tmcconne@sedona
Copyright (C) 2000, 2006-2014, 2016 Free Software Foundation, Inc. </text>




##-------------------------
## License Information
##-------------------------

leimh@pum$</text>




##-------------------------
## License Information
##-------------------------
```

**Comment by deveaud-m** [comment]
Thanks @leimaohui for reporting this issue!

Are you trying to generate SPDX reports for the same upload using different fossology-python clients running in parallel and getting different results?

Each fossology-python client uses a unique HTTP session with the Fossology host, therefore concurrency issues shall not happen on client level, or you're using multi-threading in your program which is currently not supported in this library.

Could please elaborate a bit more on your use-case?

**Comment by leimaohui** [comment]
My case is invoking fossology-python in Yocto, and run multi tasks to get spdx files for different OSS. I found that some spdx files don't have "License Information". 
Even, I [Download SPDX2 tag/value report] from "Show Jobs " page manually, the "License Information" is still empty. But if I "[Export SPDX tag:value]" on “Browse ” page, the "License Information" will always appears.
So, I guess, is it because fossology server is too busy to get "License Information"？ I wonder 
 if fossology-python can check the "License Information" field?

**Comment by deveaud-m** [comment]
I guess this is a bug in Fossology then, exposed SPXD (and maybe other) reports seem not finished or corrupt.

If the clearing information didn't change, subsequent report generations shall contain the same information. It seems to me that there is some conflict when running report generation in parallel on the Fossology server.

The REST API is just  a different representation of the Fossology assets and it seems to work correctly in your case: report downloaded on the web UI is equal to the report available on the server side. We need to ensure that the report generation is working as expected in the backend.

@ag4ums what do you think? Shall we move this issue to Fossology?

**Comment by GMishx** [comment]
Hello @leimaohui,

The SPDX report might be empty if the other scans have not been finished. Did you check if the jobs finished before generating the SPDX report? As you've mentioned, exporting the report from UI works.

**Comment by leimaohui** [comment]
Hi @GMishx，
I think the jobs before get report have been finished, because schedule_jobs returns normally.
Do you mean I must submit jobs one by one? I have more than  800+ OSS to scan. If scan them one by one, it will spend too much time.
I just notice sometimes "License Information" is empty. Do you mean the other field such as "File Information" maybe not complete too? It's different to confirm, there are too many files in an OSS.
Is there any method to know whether the spdx file report finishes normally?

**Comment by GMishx** [comment]
You do not need to do them one by one, but the `schedule_jobs` just pushes the jobs to FOSSology scheduler. The jobs will be scheduled as per the condition on FOSSology server and thus you might submit the report job before scanning even finishes.

You should use `detail_job()` to poll the server and check if the scans have finished before requesting report.

The workflow of FOSSology works like:
1. Upload a file
1. Schedule the scan
1. Wait for the scans to finish
1. Download the report.

**Comment by deveaud-m** [comment]
Thanks @GMishx for clarifying Fossology workflow!

Indeed `detail_job` will give you information about the status of jobs scheduled for a particular upload. Just ensure that all required jobs are `Completed` until generating and downloading a report.

I just remark that this library doesn't support listing jobs for a particular `upload_id` (as offered by the API). I'll create a separate issue and implement this.

**Comment by GMishx** [comment]
Just a suggestion @deveaud-m, does it sound good to have a waiter object which can be used to hold the thread in wait until the `detail_job` or new function returns `Completed` or raise an exception after a timeout or failure?

**Comment by deveaud-m** [comment]
This could be a method parameter to support different workflow, sequential and parallelized.

Thanks for the suggestion.

**Comment by deveaud-m** [comment]
@leimaohui I just released version 0.1.0 with new features relevant for your use-case:
* [schedule_jobs](https://fossology.github.io/fossology-python/jobs.html#fossology.jobs.Jobs.schedule_jobs) now accepts `wait` and `timeout` options
* [detail_job](https://fossology.github.io/fossology-python/jobs.html#fossology.jobs.Jobs.detail_job) too
* [list_jobs](https://fossology.github.io/fossology-python/jobs.html#fossology.jobs.Jobs.list_jobs) can limit the search to jobs related to a specific upload

I hope this will help!

**Comment by leimaohui** [comment]
I'll try the new feature. Thank you very much.

**Comment by deveaud-m** [comment]
@leimaohui I'll close this issue now since the required feature is available since version 0.1.0.

Don't hesitate to re-open if you're still missing something!
